aida/tutorial04/tutorial04.py

```diff
@@ -0,0 +1,145 @@
+        best_edge = {}
+        best_score['0 <s>'] = 0
+        best_edge['0 <s>'] = None
+        for i in range(len(words)):
+            for prev in self.possible_tags:
+                for next in self.possible_tags:
+                    previd_prevtag = f'{i} {prev}'
+                    prevtag_nexttag = f'{prev} {next}'
+                    nexttag_prevword = f'{next} {words[i]}'
+
+                    if previd_prevtag in best_score and prevtag_nexttag in self.transition:
+                        score = best_score[previd_prevtag] - log(self.prob_trans[prevtag_nexttag], 2)
+                        prob_e = unknown_rate / N
+                        if nexttag_prevword in self.prob_emit:
+                            prob_e += (1 - unknown_rate) * self.prob_emit[nexttag_prevword]
+                        score += -log(prob_e, 2)
+
+                        nextid_nexttag = f'{i+1} {next}'
+                        if nextid_nexttag not in best_score or score < best_score[nextid_nexttag]:
+                            best_score[nextid_nexttag] = score
+                            best_edge[nextid_nexttag] = previd_prevtag
+        #consider EOS, </s>
+        for tag in self.possible_tags:
+            secondendid_tag = f'{len(words)} {tag}'
+            tag_eos = f'{tag} </s>'
+            if secondendid_tag in best_score and tag_eos in self.prob_trans:
+                score = best_score[secondendid_tag] - log(self.prob_trans[tag_eos], 2)
+                endid_eos = f'{len(words)+1} </s>'
+                if endid_eos not in best_score or score < best_score[endid_eos]:
+                    best_score[endid_eos] = score
+                    best_edge[endid_eos] = secondendid_tag
+
+        return best_score, best_edge, endid_eos
+
+    def backward(self, best_score, best_edge, endid_eos):
+        """ viterbi backword
+        :param best_score: dict, best scores[id_tag]: score
+        :param best_edge: dict, best edge[nextid_nexttag]: previd_prevtag
+
+        :return join(tags): predicted tags
+        """
+        tags = []
+        next_edge = best_edge[endid_eos]
+        while next_edge != '0 <s>':
+            position, tag = next_edge.split()
+            tags.append(tag)
+            next_edge = best_edge[next_edge]
+        tags = tags[::-1]
+        return print(' '.join(tags))
+
+    def test(self, file_path):
+        """ estimate tags
+        :param file_path: the path of test data
+        
+        :return tags(from 'backward'): estimated tags for each words
+        """
+        with open(file_path) as fp:
+            for line in fp:
+                words = line.strip().split()
+                #print(words)
+                best_score, best_edge, endid_eos = self.forward(words)
+                self.backward(best_score, best_edge, endid_eos)
+        return
+
+if __name__ == '__main__':
+    #train_file_path = '../test/05-train-input.txt'
+    train_file_path = '../data/wiki-en-train.norm_pos'
+    HMM = HMM()
```

Comment:
```suggestion
    hmm = HMM()
```

`HMM = HMM()` と書いてしまうと，`HMM` というオブジェクトは

- この文より前では `Class`
- この文より後では `HMM` だったもののインスタンス

になってしまいます

つまり，以降 `HMM()` ができなくなってしまうので，クラスの名前と異なる変数名にするのがおすすめです

- クラスは，最初大文字 + 大文字区切り
  - e.g. `MyFavoriteClass`
- 変数は，全小文字 + アンダースコア区切り
  - e.g. `my_favorite_instance`

という命名規則はよくできているのかもしれませんね

Reply:
ご指摘ありがとうございます！
Class の名前とインスタンスの名前が衝突していますね、修正します。